### PR TITLE
23 implement saving player highest score in .txt file

### PR DESCRIPTION
**Pull request by Svetlanazub**
Implemented feature

### CODE REVIEW

**Comment by marcin-ogorzalek-EPAM** [comment]
You have to bump up the project's version in pom.
Also, there is no link to the issue in the description, especially one closing the issue on merge.

File: pom.xml [hunk 1]
```diff
@@ -17,6 +17,7 @@
         <version.plugin.maven.enforcer>3.0.0-M2</version.plugin.maven.enforcer>
         <version.plugin.maven.jar>3.2.0</version.plugin.maven.jar>
         <version.testng>7.4.0</version.testng>
+        <maven.assembly.plugin.version>3.3.0</maven.assembly.plugin.version>
```
Comment (marcin-ogorzalek-EPAM) [inline pom.xml:20]: The name of this variable breaks an already established naming convention.

Comment (marcin-ogorzalek-EPAM) [reply pom.xml:20]: Why after removal `jar` plugin variable for it is still present?

File: pom.xml [hunk 2]
```diff
@@ -26,22 +27,40 @@
             <version>${version.testng}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.8.7</version>
```
Comment (marcin-ogorzalek-EPAM) [inline pom.xml:33]: why not in `properties`?

File: src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * These tests check:
+ * <p><ul>
+ * <li>1. Return appropriate message when JSON file doesn't exist
+ * <li>2. Generating 3 chars
+ * <li>3. toString method for printing 25 lines
+ * <li>4. sorting scores from max to min
+ * </ul><p>
+ *
+ * @author Svetlana_Zubkova
+ */
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:18]: By convention, tests shouldn't have javadocs. Unless there are some trickiness in the tests themselves.

File: src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Writes players' scores in JSON file
+ * Prints 25 maximum score in order from the highest to lowest before current player's score
+ *
+ * @author Svetlana_Zubkova
+ */
+class HighestStoredScores {
+    private final List<PlayerScore> scoresList;
+    private final static int MAX_AMOUNT_OF_ENTRIES = 25;
+    private final String path;
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java:15]: Reorder those. Constant should be first.

File: src/main/java/com/epam/prejap/teatrees/PlayerScore.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.epam.prejap.teatrees;
+
+/**
+ * Creates record with player's names and scores
+ * @author Svetlana_Zubkova
+ */
+class PlayerScore implements Comparable<PlayerScore> {
+    private final String name;
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/PlayerScore.java:8]: Wrong intentation.

File: src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Reads previous highest scores from JSON file and writes new one to it
+ * @author Svetlana_Zubkova
+ */
+class SavedScoreIOUtility {
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java:17]: Why this class was created as a helper?

File: src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    HighestStoredScores(String filePath) {
+        if (filePath != null) {
+            path = filePath;
+            scoresList = SavedScoreIOUtility.readSavedScore(filePath);
+        } else {
+            throw new IllegalArgumentException("File don't find!");
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java:22]: Are we catching this somehow or are we just crashing the app? I didn't spot try-catch for it in main or somewhere else.

File: src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            InputStream inputStream = getInputStream(name);
+            InputStreamReader streamReader = new InputStreamReader(inputStream);
+            return new BufferedReader(streamReader);
+        }
+
+        private static InputStream getInputStream(String name) {
+            ClassLoader classLoader = SavedScoreIOUtility.class.getClassLoader();
+            InputStream inputStream = classLoader.getResourceAsStream(name);
+                if (inputStream == null) {
+                    throw new IllegalArgumentException("You are first player!");
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java:42]: Are we catching this somehow or are we just crashing the app? I didn't spot try-catch for it in main or somewhere else.

File: src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Generates current player's name from 3 characters
+     *
+     * @return String
+     */
+    static String generateName() {
+        final String alphabet = "ABCDEFGHIJKLMNOPRQSTUXYZ";
+        int n = alphabet.length();
+        Random r = new Random();
+        StringBuilder input = new StringBuilder();
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/HighestStoredScores.java:55]: Why name `input`?

File: src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+
+    @Test(expectedExceptions = IllegalArgumentException.class,
+            expectedExceptionsMessageRegExp = "File don't find!")
+    public void shouldReturnAppropriateMessageWhenNullMapInput() {
+        new HighestStoredScores(null);
+    }
+
+
+    @BeforeClass(groups = {"IntegrationHighestStoredScoresTests"})
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:31]: I like an establishment of tests' groups

File: src/main/java/com/epam/prejap/teatrees/PlayerScore.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package com.epam.prejap.teatrees;
+
+/**
+ * Creates record with player's names and scores
+ * @author Svetlana_Zubkova
+ */
+class PlayerScore implements Comparable<PlayerScore> {
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/PlayerScore.java:7]: Why not record?

File: src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+ * <p><ul>
+ * <li>1. Return appropriate message when JSON file doesn't exist
+ * <li>2. Generating 3 chars
+ * <li>3. toString method for printing 25 lines
+ * <li>4. sorting scores from max to min
+ * </ul><p>
+ *
+ * @author Svetlana_Zubkova
+ */
+public class HighestStoredScoresTest {
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:19]: As this read file, I think that this is an integration test.

Comment (marcin-ogorzalek-EPAM) [reply src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:19]: or rather we have a mix here. Divide those into two classes. One with unit tests and one with integrational.

File: src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    public void shouldGenerateThreeCharsName() {
+        String testName = HighestStoredScores.generateName();
+        assertEquals(testName.length(), 3);
+    }
+
+    @Test(groups = {"IntegrationHighestStoredScoresTests"})
+    public void shouldPrint25Lines() {
+        String[] sizeOfPrinting = storedScore.toString().split("/n");
+        if (sizeOfPrinting.length > 25) {
+            fail();
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:48]: `assertFalse` may be used instead

File: src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Test(expectedExceptions = IllegalArgumentException.class,
+            expectedExceptionsMessageRegExp = "File don't find!")
+    public void shouldReturnAppropriateMessageWhenNullMapInput() {
+        new HighestStoredScores(null);
+    }
+
+
+    @BeforeClass(groups = {"IntegrationHighestStoredScoresTests"})
+    public void setUp() {
```
Comment (marcin-ogorzalek-EPAM) [inline src/test/java/com/epam/prejap/teatrees/HighestStoredScoresTest.java:32]: for readability this should be place before the first test

File: src/main/java/com/epam/prejap/teatrees/TeaTrees.java [hunk 15]
```diff
@@ -42,18 +42,23 @@ public Score play() {
     }
 
     public static void main(String[] args) {
+
         int rows = 10;
         int cols = 20;
-        int delay = 500;
-
+        int delay = 5;
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/TeaTrees.java:48]: why the delay has decreased?

File: src/main/java/com/epam/prejap/teatrees/TeaTrees.java [hunk 16]
```diff
@@ -42,18 +42,23 @@ public Score play() {
...
+        final String nameJSONFile = "score_list.json";
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/TeaTrees.java:49]:
```suggestion
        final String scoreListFileName = "score_list.json";
```

JSONFile name is meaningless, doesn't tell what this is really about.

File: src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    static List<PlayerScore> readSavedScore(String name) {
+        var results = new ArrayList<PlayerScore>();
+            BufferedReader reader = getBufferedReader(name);
+            results = new Gson().fromJson(reader, new TypeToken<List<PlayerScore>>() {
+            }.getType());
+        Collections.sort(results);
+            return results;
+        }
+
+        private static BufferedReader getBufferedReader(String name) {
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java:32]: Why `BufferReader` as a return type. Is this multithread?

File: src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+            InputStream inputStream = getInputStream(name);
+            InputStreamReader streamReader = new InputStreamReader(inputStream);
+            return new BufferedReader(streamReader);
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prejap/teatrees/SavedScoreIOUtility.java:35]: I'm not a fan of this method. `BufferedReader` and `InputStream` should be closed at the end of use. What is more those should be close by try-with-resources.

**Review by marcin-ogorzalek-EPAM — CHANGES_REQUESTED** [review]
A couple of things to improve here. 
Remember about syncing your branch with the `master` on the main repo. 
Bump up the project's version in `pom.xml`. Update `testng.xml`
And link this PR to the proper issue via `Closes #23` in the description

**Comment by marcin-ogorzalek-EPAM** [comment]
As I had been informed, this probably won't be developed any further in the closest future. Probably for closing @pkierat-epam

**Comment by Svetlanazub** [comment]
@pkierat-epam I'll be able to finish the task in some next days so I close it and someone can do it.

**Pull request closed without merging**